zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java

```diff
@@ -417,19 +436,42 @@ synchronized private void stop()
             } catch (IllegalStateException e) {
                 // This can happen if the VM is already shutting down
                 LOG.debug("Failed to remove shutdown hook", e);
+                throw e;
             }
 
             stopped = true;
             LOG.info("Done shutting down");
         }
     }
 
-    static Map<SocketAddress, ChannelInitializer<?>> convertPortMap(
+    /**
+     * Keys should be a short string usable in metrics.
+     */
+    public static final AttributeKey<Attrs> CONN_DIMENSIONS = AttributeKey.newInstance("zuulconndimensions");
+
+    private final class NewConnHandler extends ChannelInboundHandlerAdapter {
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+            Long now = System.nanoTime();
+            final Channel child = (Channel) msg;
+            child.attr(CONN_DIMENSIONS).set(Attrs.newInstance());
+            ConnTimer timer = ConnTimer.install(child, registry, registry.createId("zuul.conn.client.timing"));
+            timer.record(now, "ACCEPT");
+            ConnCounter.install(child, registry, registry.createId("zuul.conn.client.current"));
+            super.channelRead(ctx, msg);
+        }
+    }
+
+    static Map<NamedSocketAddress, ChannelInitializer<?>> convertPortMap(
             Map<Integer, ChannelInitializer<?>> portsToChannelInitializers) {
-        Map<SocketAddress, ChannelInitializer<?>> addrsToInitializers =
+        Map<NamedSocketAddress, ChannelInitializer<?>> addrsToInitializers =
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The `HashMap` constructor rehashes your hash table after 75% of all keys have been added to the table. To improve performance and prevent rehashing, use the Guava `newHashMapWithExpectedSize()` method. See the following diffs on GitHub that make this code change: [diff1](https://github.com/thefreebit/bullfrog/commit/d4d0b55af8a0eb925f039ef78d137565e1ae574b#diff-a24080844c93ebee0f466f384d2b3599cbfd502623597fc30e3364bc3fad5fedL119), [diff2](https://github.com/outbrain/ob1k/commit/5183bd681f4e9a8b67289314359b1211eb764899#diff-1891c9f8b2921c25e9406be78b9688662c7cd63b57300817fd288ed2ec0e3eeaL282)

Learn more: [HashMap constructor in the Java Platform Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#HashMap-int-) and the [newHashMapWithExpectedSize in the Guava: Google Core Libraries for Java 23.0 API](https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Maps.html#newHashMapWithExpectedSize-int-)